File: src/nifs/drawing.zig
const std = @import("std");
const assert = std.debug.assert;
const e = @import("../erl_nif.zig");
const rl = @import("../raylib.zig");

const core = @import("../core.zig");

pub const exported_nifs = [_]e.ErlNifFunc{
    // Drawing
    .{ .name = "clear_background", .arity = 1, .fptr = core.nif_wrapper(nif_clear_background), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_drawing", .arity = 0, .fptr = core.nif_wrapper(nif_begin_drawing), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_drawing", .arity = 0, .fptr = core.nif_wrapper(nif_end_drawing), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_mode_2d", .arity = 1, .fptr = core.nif_wrapper(nif_begin_mode_2d), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_mode_2d", .arity = 0, .fptr = core.nif_wrapper(nif_end_mode_2d), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_mode_3d", .arity = 1, .fptr = core.nif_wrapper(nif_begin_mode_3d), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_mode_3d", .arity = 0, .fptr = core.nif_wrapper(nif_end_mode_3d), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_texture_mode", .arity = 1, .fptr = core.nif_wrapper(nif_begin_texture_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_texture_mode", .arity = 0, .fptr = core.nif_wrapper(nif_end_texture_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_shader_mode", .arity = 1, .fptr = core.nif_wrapper(nif_begin_shader_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_shader_mode", .arity = 0, .fptr = core.nif_wrapper(nif_end_shader_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_blend_mode", .arity = 1, .fptr = core.nif_wrapper(nif_begin_blend_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_blend_mode", .arity = 0, .fptr = core.nif_wrapper(nif_end_blend_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_scissor_mode", .arity = 1, .fptr = core.nif_wrapper(nif_begin_scissor_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_scissor_mode", .arity = 4, .fptr = core.nif_wrapper(nif_begin_scissor_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_scissor_mode", .arity = 0, .fptr = core.nif_wrapper(nif_end_scissor_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "begin_vr_stereo_mode", .arity = 1, .fptr = core.nif_wrapper(nif_begin_vr_stereo_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "end_vr_stereo_mode", .arity = 0, .fptr = core.nif_wrapper(nif_end_vr_stereo_mode), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
};

///////////////
//  Drawing  //
///////////////

/// Set background color (framebuffer clear color)
///
/// raylib.h
/// RLAPI void ClearBackground(Color color);
fn nif_clear_background(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const arg_color = core.Argument(core.Color).get(env, argv[0]) catch {
        return error.invalid_argument_color;
    };
    defer arg_color.free();
    const color = arg_color.data;

    // Function

    rl.ClearBackground(color);

    // Return

    return core.Atom.make(env, "ok");
}

/// Setup canvas (framebuffer) to start drawing
///
/// raylib.h
/// RLAPI void BeginDrawing(void);
fn nif_begin_drawing(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.BeginDrawing();

    // Return

    return core.Atom.make(env, "ok");
}

/// End canvas drawing and swap buffers (double buffering)
///
/// raylib.h
/// RLAPI void EndDrawing(void);
fn nif_end_drawing(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndDrawing();

    // Return

    return core.Atom.make(env, "ok");
}

/// Begin 2D mode with custom camera (2D)
///
/// raylib.h
/// RLAPI void BeginMode2D(Camera2D camera);
fn nif_begin_mode_2d(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const arg_camera = core.Argument(core.Camera2D).get(env, argv[0]) catch {
        return error.invalid_argument_camera;
    };
    defer arg_camera.free();
    const camera = arg_camera.data;

    // Function

    rl.BeginMode2D(camera);

    // Return

    return core.Atom.make(env, "ok");
}

/// Ends 2D mode with custom camera
///
/// raylib.h
/// RLAPI void EndMode2D(void);
fn nif_end_mode_2d(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndMode2D();

    // Return

    return core.Atom.make(env, "ok");
}

/// Begin 3D mode with custom camera (3D)
///
/// raylib.h
/// RLAPI void BeginMode3D(Camera3D camera);
fn nif_begin_mode_3d(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const arg_camera = core.Argument(core.Camera3D).get(env, argv[0]) catch {
        return error.invalid_argument_camera;
    };
    defer arg_camera.free();
    const camera = arg_camera.data;

    // Function

    rl.BeginMode3D(camera);

    // Return

    return core.Atom.make(env, "ok");
}

/// Ends 3D mode and returns to default 2D orthographic mode
///
/// raylib.h
/// RLAPI void EndMode3D(void);
fn nif_end_mode_3d(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndMode3D();

    // Return

    return core.Atom.make(env, "ok");
}

/// Begin drawing to render texture
///
/// raylib.h
/// RLAPI void BeginTextureMode(RenderTexture2D target);
fn nif_begin_texture_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const arg_target = core.Argument(core.RenderTexture2D).get(env, argv[0]) catch {
        return error.invalid_argument_target;
    };
    defer arg_target.free();
    const target = arg_target.data;

    // Function

    rl.BeginTextureMode(target);

    // Return

    return core.Atom.make(env, "ok");
}

/// Ends drawing to render texture
///
/// raylib.h
/// RLAPI void EndTextureMode(void);
fn nif_end_texture_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndTextureMode();

    // Return

    return core.Atom.make(env, "ok");
}

/// Begin custom shader drawing
///
/// raylib.h
/// RLAPI void BeginShaderMode(Shader shader);
fn nif_begin_shader_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const arg_shader = core.Argument(core.Shader).get(env, argv[0]) catch {
        return error.invalid_argument_shader;
    };
    defer arg_shader.free();
    const shader = arg_shader.data;

    // Function

    rl.BeginShaderMode(shader);

    // Return

    return core.Atom.make(env, "ok");
}

/// End custom shader drawing (use default shader)
///
/// raylib.h
/// RLAPI void EndShaderMode(void);
fn nif_end_shader_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndShaderMode();

    // Return

    return core.Atom.make(env, "ok");
}

/// Begin blending mode (alpha, additive, multiplied, subtract, custom)
///
/// raylib.h
/// RLAPI void BeginBlendMode(int mode);
fn nif_begin_blend_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const mode = core.Int.get(env, argv[0]) catch {
        return error.invalid_argument_mode;
    };

    // Function

    rl.BeginBlendMode(mode);

    // Return

    return core.Atom.make(env, "ok");
}

/// End blending mode (reset to default: alpha blending)
///
/// raylib.h
/// RLAPI void EndBlendMode(void);
fn nif_end_blend_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndBlendMode();

    // Return

    return core.Atom.make(env, "ok");
}

/// Begin scissor mode (define screen area for following drawing)
///
/// raylib.h
/// RLAPI void BeginScissorMode(int x, int y, int width, int height);
fn nif_begin_scissor_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1 or argc == 4);

    // Arguments

    var x: c_int = undefined;
    var y: c_int = undefined;
    var width: c_int = undefined;
    var height: c_int = undefined;

    if (argc == 1) {
        const arg_rec = core.Argument(core.Rectangle).get(env, argv[0]) catch {
            return error.invalid_argument_rec;
        };
        defer arg_rec.free();
        const rec = arg_rec.data;

        x = @intFromFloat(rec.x);
        y = @intFromFloat(rec.y);
        width = @intFromFloat(rec.width);
        height = @intFromFloat(rec.height);
    } else {
        x = core.Int.get(env, argv[0]) catch {
            return error.invalid_argument_x;
        };

        y = core.Int.get(env, argv[1]) catch {
            return error.invalid_argument_y;
        };

        width = core.Int.get(env, argv[2]) catch {
            return error.invalid_argument_width;
        };

        height = core.Int.get(env, argv[3]) catch {
            return error.invalid_argument_height;
        };
    }

    // Function

    rl.BeginScissorMode(x, y, width, height);

    // Return

    return core.Atom.make(env, "ok");
}

/// End scissor mode
///
/// raylib.h
/// RLAPI void EndScissorMode(void);
fn nif_end_scissor_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndScissorMode();

    // Return

    return core.Atom.make(env, "ok");
}

/// Begin stereo rendering (requires VR simulator)
///
/// raylib.h
/// RLAPI void BeginVrStereoMode(VrStereoConfig config);
fn nif_begin_vr_stereo_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 1);

    // Arguments

    const arg_config = core.Argument(core.VrStereoConfig).get(env, argv[0]) catch {
        return error.invalid_argument_config;
    };
    defer arg_config.free();
    const config = arg_config.data;

    // Function

    rl.BeginVrStereoMode(config);

    // Return

    return core.Atom.make(env, "ok");
}

/// End stereo rendering (requires VR simulator)
///
/// raylib.h
/// RLAPI void EndVrStereoMode(void);
fn nif_end_vr_stereo_mode(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    rl.EndVrStereoMode();

    // Return

    return core.Atom.make(env, "ok");
}

File: lib/zexray/nif/drawing.ex
defmodule Zexray.NIF.Drawing do
  @moduledoc false

  defmacro __using__(_opts) do
    quote do
      @nifs_drawing [
        # Drawing
        clear_background: 1,
        begin_drawing: 0,
        end_drawing: 0,
        begin_mode_2d: 1,
        end_mode_2d: 0,
        begin_mode_3d: 1,
        end_mode_3d: 0,
        begin_texture_mode: 1,
        end_texture_mode: 0,
        begin_shader_mode: 1,
        end_shader_mode: 0,
        begin_blend_mode: 1,
        end_blend_mode: 0,
        begin_scissor_mode: 4,
        end_scissor_mode: 0,
        begin_vr_stereo_mode: 1,
        end_vr_stereo_mode: 0
      ]

      #############
      #  Drawing  #
      #############

      @doc """
      Set background color (framebuffer clear color)

      ```c
      // raylib.h
      RLAPI void ClearBackground(Color color);
      ```
      """
      @doc group: :drawing
      @spec clear_background(color :: map | reference) :: :ok
      def clear_background(_color), do: :erlang.nif_error(:undef)

      @doc """
      Setup canvas (framebuffer) to start drawing

      ```c
      // raylib.h
      RLAPI void BeginDrawing(void);
      ```
      """
      @doc group: :drawing
      @spec begin_drawing() :: :ok
      def begin_drawing(), do: :erlang.nif_error(:undef)

      @doc """
      End canvas drawing and swap buffers (double buffering)

      ```c
      // raylib.h
      RLAPI void EndDrawing(void);
      ```
      """
      @doc group: :drawing
      @spec end_drawing() :: :ok
      def end_drawing(), do: :erlang.nif_error(:undef)

      @doc """
      Begin 2D mode with custom camera (2D)

      ```c
      // raylib.h
      RLAPI void BeginMode2D(Camera2D camera);
      ```
      """
      @doc group: :drawing
      @spec begin_mode_2d(camera :: map | reference) :: :ok
      def begin_mode_2d(_camera), do: :erlang.nif_error(:undef)

      @doc """
      Ends 2D mode with custom camera

      ```c
      // raylib.h
      RLAPI void EndMode2D(void);
      ```
      """
      @doc group: :drawing
      @spec end_mode_2d() :: :ok
      def end_mode_2d(), do: :erlang.nif_error(:undef)

      @doc """
      Begin 3D mode with custom camera (3D)

      ```c
      // raylib.h
      RLAPI void BeginMode3D(Camera3D camera);
      ```
      """
      @doc group: :drawing
      @spec begin_mode_3d(camera :: map | reference) :: :ok
      def begin_mode_3d(_camera), do: :erlang.nif_error(:undef)

      @doc """
      Ends 3D mode and returns to default 2D orthographic mode

      ```c
      // raylib.h
      RLAPI void EndMode3D(void);
      ```
      """
      @doc group: :drawing
      @spec end_mode_3d() :: :ok
      def end_mode_3d(), do: :erlang.nif_error(:undef)

      @doc """
      Begin drawing to render texture

      ```c
      // raylib.h
      RLAPI void BeginTextureMode(RenderTexture2D target);
      ```
      """
      @doc group: :drawing
      @spec begin_texture_mode(target :: map | reference) :: :ok
      def begin_texture_mode(_target), do: :erlang.nif_error(:undef)

      @doc """
      Ends drawing to render texture

      ```c
      // raylib.h
      RLAPI void EndTextureMode(void);
      ```
      """
      @doc group: :drawing
      @spec end_texture_mode() :: :ok
      def end_texture_mode(), do: :erlang.nif_error(:undef)

      @doc """
      Begin custom shader drawing

      ```c
      // raylib.h
      RLAPI void BeginShaderMode(Shader shader);
      ```
      """
      @doc group: :drawing
      @spec begin_shader_mode(shader :: map | reference) :: :ok
      def begin_shader_mode(_shader), do: :erlang.nif_error(:undef)

      @doc """
      End custom shader drawing (use default shader)

      ```c
      // raylib.h
      RLAPI void EndShaderMode(void);
      ```
      """
      @doc group: :drawing
      @spec end_shader_mode() :: :ok
      def end_shader_mode(), do: :erlang.nif_error(:undef)

      @doc """
      Begin blending mode (alpha, additive, multiplied, subtract, custom)

      ```c
      // raylib.h
      RLAPI void BeginBlendMode(int mode);
      ```
      """
      @doc group: :drawing
      @spec begin_blend_mode(mode :: integer) :: :ok
      def begin_blend_mode(_mode), do: :erlang.nif_error(:undef)

      @doc """
      End blending mode (reset to default: alpha blending)

      ```c
      // raylib.h
      RLAPI void EndBlendMode(void);
      ```
      """
      @doc group: :drawing
      @spec end_blend_mode() :: :ok
      def end_blend_mode(), do: :erlang.nif_error(:undef)

      @doc """
      Begin scissor mode (define screen area for following drawing)

      ```c
      // raylib.h
      RLAPI void BeginScissorMode(int x, int y, int width, int height);
      ```
      """
      @doc group: :drawing
      @spec begin_scissor_mode(
              x :: integer,
              y :: integer,
              width :: integer,
              height :: integer
            ) :: :ok
      def begin_scissor_mode(
            _x,
            _y,
            _width,
            _height
          ),
          do: :erlang.nif_error(:undef)

      @doc """
      End scissor mode

      ```c
      // raylib.h
      RLAPI void EndScissorMode(void);
      ```
      """
      @doc group: :drawing
      @spec end_scissor_mode() :: :ok
      def end_scissor_mode(), do: :erlang.nif_error(:undef)

      @doc """
      Begin stereo rendering (requires VR simulator)

      ```c
      // raylib.h
      RLAPI void BeginVrStereoMode(VrStereoConfig config);
      ```
      """
      @doc group: :drawing
      @spec begin_vr_stereo_mode(config :: map | reference) :: :ok
      def begin_vr_stereo_mode(_config), do: :erlang.nif_error(:undef)

      @doc """
      End stereo rendering (requires VR simulator)

      ```c
      // raylib.h
      RLAPI void EndVrStereoMode(void);
      ```
      """
      @doc group: :drawing
      @spec end_vr_stereo_mode() :: :ok
      def end_vr_stereo_mode(), do: :erlang.nif_error(:undef)
    end
  end
end
